src/libcore/cmp.rs

```diff
@@ -462,6 +462,40 @@ impl<T: Ord> Ord for Reverse<T> {
     }
 }
 
+/// A wrapper newtype for providing a total order on a type that's normally partial.
+///
+/// This is most commonly used for floating-point:
+///
+/// ```rust
+/// #![feature(float_total_cmp)]
+///
+/// use std::cmp::Total;
+///
+/// assert!(-0.0 == 0.0);
+/// assert!(Total(-0.0) < Total(0.0));
+/// assert_eq!(std::f32::NAN.partial_cmp(&0.0), None);
+/// assert_eq!(Total(std::f32::NAN).partial_cmp(&Total(0.0)), Some(std::cmp::Ordering::Greater));
+///
+/// let mut a = [3.0, 1.0, 2.0];
+/// // a.sort(); // ERROR, because floats are !Ord
+/// a.sort_by_key(|x| std::cmp::Total(*x)); // But this works!
+/// assert_eq!(a, [1.0, 2.0, 3.0]);
+///
+/// // By using `Total`, the struct can derive `Eq` and `Ord`.
+/// #[derive(PartialEq, Eq, PartialOrd, Ord)]
+/// struct MyData {
+///     foo: Total<f32>,
+///     bar: Total<f64>,
+/// }
+/// ```
+///
+/// It can also be used to provide both partial and total order implementations
+/// for an enum, if orders between variant don't make sense conceptually but
+/// are still desired for use as a `BTreeMap` key or similar.
+#[derive(Debug, Copy, Clone)]
+#[unstable(feature = "float_total_cmp", issue = "55339")]
+pub struct Total<T>(#[unstable(feature = "float_total_cmp", issue = "55339")] pub T);
+
```

Comment:
The `PartialOrd`/`Ord` "API" makes the assumption that only a single three-way comparator makes sense for any given type. This API covers the case in which more than one comparator makes sense, and the one implemented for the type is not total.

This feel like an extremely niche case for such a general API in `core::cmp`. For example, are we going to cover the opposite case, e.g., by adding `Partial<T>` as well? What about types with two total orders, where users would like to use the non-default one: are we going to add a `Total2<T>` to `core::cmp` to cover that as well?

Floats are ubiquitous, and it makes sense to make them easier to use. I think we could add `[f32, f64]::total_order(self) -> TotalFloat<Self>` methods to the floating-point types, and a `TotalFloat` type wrapper to `core::float` (or somewhere else) that implements the exact same semantics as a float, but where the default ordering is total instead of partial. That's a smaller and more focused problem to solve.

In `std::cmp`, either we should try to solve the underlying problem, or provide better duct-tape, e.g., `Total<T, Fn(Self, Self)-> Ordering>` and `Partial<T, Fn(Self,Self)->Option<Ordering>>`, such that one can write `Total<f32, f32::total_cmp>` to select the orderings or similar.